pkg/chunk/chunk_store.go

```diff
@@ -262,40 +267,52 @@ outer:
 		}
 
 		// Apply metric name matcher
-		if metricNameMatcher != nil && !metricNameMatcher.Match(metric[metricNameMatcher.Name]) {
+		if metricNameMatcher != nil && !metricNameMatcher.Matches(string(metric[model.LabelName(metricNameMatcher.Name)])) {
 			continue outer
 		}
 
 		// Apply matchers
 		for _, matcher := range allMatchers {
-			if !matcher.Match(metric[matcher.Name]) {
+			if !matcher.Matches(string(metric[model.LabelName(matcher.Name)])) {
 				continue outer
 			}
 		}
 
-		orgID, err := user.ExtractOrgID(ctx)
-		if err != nil {
-			return nil, err
+		// TODO(prom2): Does this contraption over-match?
```

Comment:
How would we test this?

Reply:
This is something I still want to look at closer before merging. Mainly I extracted this code from the lazy iterators and moved it to here (for eager loading). I just need to reason through it all one more time before giving a better answer :)

Reply:
So to elaborate:

The question I have around this code is not related to this PR really, as I just moved it here to be eager-loading instead of lazy-loading. The same question applies to the current `master`.

Somehow it seems to actually select only the correct series, although if I run it in my head there should be a bug: the code gets all series where the metric name matcher applies, then filters them down to only the series where the other matchers also apply. From those matching series, it builds a set of new matchers that are then used to look up the final series chunks. But those matchers should then match too many series in some cases.

So e.g. I thought this would over-match in the following scenario:

- TSDB has both `foo{a="b"}` and `foo{a="b", c="d"}`
- Query is for `{__name__=~"foo", c!="d"}`
  - (correctly this should return `foo{a="b"}`, but not `foo{a="b", c="d"})`

What I would expect to happen in the code is then:

1. it finds and matches both `foo` series in the metric name index
2. it filters out the `foo{a="b", c="d"}` series because it doesn't match the `c!="d"` matcher
3. it builds new matchers out of the remaining series's labels (`foo{a="b"}`)
4. it uses those new matchers to look up all series that match them, which should *also* match `foo{a="b", c="d"}`, since it is a superset of labels

But somehow, it does not return this wrong series though and the query result is correct. But why? Do you see what I'm missing?

Reply:
So of course I couldn't just sleep with this mystery unsolved. Many debug Printf's later, I found out the following:

- The returned series were only correct because they came from the ingesters only, and not from the chunk store.
- The chunk store actually didn't select *any* data for the above sample scenario because of a bug I introduced (fixed in latest push) that passed in `__name__` instead of the metric name into `getMetricNameMatrix()`.
- The current `master` is actually worse: rather than selecting too many series, the lazy iterator just crashes with a `nil pointer dereference` in the above scenario, as soon as there are relevant chunks in S3. This is currently happening in production as well.

So: `master` is currently broken in a crashy way (queries fail completely), whereas this PR branch is broken in that it selects too many series in an edge case. That problem would also occur in `master`, if it didn't crash before that. Thus, I am almost tending to not try and fix this unrelated bug in this PR. @jml What do you think?